### PR TITLE
[RFC] Resolve neovim issue #7937

### LINKED ISSUE #7937: screen.c index out of bounds crash
- `nvim --version`:
```
:version
NVIM v0.2.3-383-gdc4a9cbe8
Build type: Debug
LuaJIT 2.0.5
Compilation: /Library/Developer/CommandLineTools/usr/bin/cc -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla
 -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -I/Users/mb013619/dev/neovim/build/config -I/Users/mb013619/dev/neovim/src -I/Users/mb013619/dev/neovim/.deps/usr/include -I/Users/mb013619/dev/neovim/.deps/usr/include -I/Users/mb013619/d
ev/neovim/.deps/usr/include -I/Users/mb013619/dev/neovim/.deps/usr/include -I/Users/mb013619/dev/neovim/.deps/usr/include -I/usr/local/opt/gettext/include -I/usr/include -I/Users/mb013619/dev/neovim/build/src/nvim/auto -I/Users/mb013619/dev/neovim/build/include
Compiled by mb013619

Features: +acl +iconv -jemalloc +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"
```
- Vim (version: ) behaves differently?
- Operating system/version:  Mac OSX 10.12.6
- Terminal name/version: tmux in iTerm2 (error also occurs when not running in tmux)
- `$TERM`:
```
:!echo $TERM
[No write since last change]

screen-256color
```

### Steps to reproduce using `nvim -u NORC`
```
$ echo Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. > lorem_ipsum
$ nvim -u NORC lorem_ipsum
gqgq
:set spell
JJJ
```
Result:
```
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna/Users/mb013619/dev/neovim/src/nvim/screen.c:3362:28: runtime error: index 301 out of bounds for type 'char_u [300]' Duis
aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint                                                                                                                       SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/mb013619/dev/neovim/src/nvim/screen.c:3362:28 in lit anim id est laborum.
~                                                                      Abort trap: 6
cwd: ~
[20180130 11:07:02|mb013619]$
```

### Actual behaviour
Using 'J' to combine several lines into a single line.  Once the line extends off the screen, this error occurs, 100% of the time:
```
neovim/src/nvim/screen.c:3362:28: runtime error: index 303 out of bounds for type 'char_u [300]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/mb013619/dev/neovim/src/nvim/screen.c:3362:28 in
Abort trap: 6
```

### Expected behaviour
There is no crash.

### PR DESCRIPTION
**Pull request by brailsmt**
Fix #7937

### CODE REVIEW

**Comment by bfredl** [comment]
Nice! Maybe we should a test for the crash in #7937 as it obviously is not a code path properly tested yet...

**Comment by mhinz** [comment]
I guess this applies to Vim as well?

**Comment by brailsmt** [comment]
I'll give writing a test a shot.

**Comment by brailsmt** [comment]
I added this test, but it is not a good one.  When I remove the fix in screen.c, the test does not cause the crash.  I could use some help to improve it such that it actually triggers the condition without the fix in place.

File: test/functional/ui/spell_spec.lua [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+local feed, insert, eq, eval, clear, expect = helpers.feed, helpers.insert, 
+  helpers.eq, helpers.eval, helpers.clear, helpers.feed_command, helpers.expect
+local Screen = require('test.functional.ui.screen')
+local feed_command = helpers.feed_command
+
+describe('combining long lines', function()
+  local screen
+  before_each(function()
+      clear()
+      screen = Screen.new(264, 4)
```
Comment (bfredl) [inline test/functional/ui/spell_spec.lua]: Maybe it triggers if the screen is narrower like 60 columns

Comment (brailsmt) [reply test/functional/ui/spell_spec.lua]: I took that from ":set columns" on my screen, which is 272 and pared it down to not look so ugly on my screen when printed.  I also tested it with 30, 78, and 80 columns.  I get the same results.  I wonder if this just isn't the best way to trigger it.

Comment (brailsmt) [reply test/functional/ui/spell_spec.lua]: Could it be that the functional testing framework is just different enough from running the full application that this bug cannot occur in the tests?

Comment (bfredl) [reply test/functional/ui/spell_spec.lua]: No, it runs full nvim instance, though some applicable option values might be different from `-u NONE`. I'll take a look.

Comment (brailsmt) [reply test/functional/ui/spell_spec.lua]: I got it to fail without the fix in place with 80 columns.  I'm not sure what is going on, but it consistently fails now.  Update imminent.

File: test/functional/ui/spell_spec.lua [hunk 2]
```diff
@@ -9,31 +9,33 @@ local feed_command = helpers.feed_command
...
-      screen:set_default_attr_ids()
   end)
 
   after_each(function()
     screen:detach()
   end)
 
   it('#7937 successfully joins lines with "set spell"', function()
-    feed_command('set spell nowrap')
+    feed_command('set nowrap spell')
```
Comment (bfredl) [inline test/functional/ui/spell_spec.lua]: Doesn't it crash also with `wrap` (the default)? I don't we should change default values if not necessary.

File: test/functional/ui/spell_spec.lua [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+local feed, insert, eq, eval, clear, expect = helpers.feed, helpers.insert, 
+  helpers.eq, helpers.eval, helpers.clear, helpers.feed_command, helpers.expect
+local Screen = require('test.functional.ui.screen')
+local feed_command = helpers.feed_command
+
+describe('combining long lines', function()
+  local screen
+
+  before_each(function()
+      clear()
```
Comment (bfredl) [inline test/functional/ui/spell_spec.lua]: Indentation, also with `screen:snapshot_util()` there should be no reason to not test highlighting.

**Comment by brailsmt** [comment]
Test run without the fix to screen.c:
```
[----------] Running tests from test/functional/ui/spell_spec.lua
[ RUN      ] combining long lines #7937 successfully joins lines with "set spell": ERR
test/functional/ui/screen.lua:307: Row 1 did not match.
Expected:
  |*^Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor i|
  |~                                                                               |
  |~                                                                               |
  |                                                                                |
Actual:
  |*                                                                                |
  |~                                                                               |
  |~                                                                               |
  |^                                                                                |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.

stack traceback:
        test/functional/ui/screen.lua:307: in function 'wait'
        test/functional/ui/screen.lua:221: in function 'expect'
        test/functional/ui/spell_spec.lua:34: in function <test/functional/ui/spell_spec.lua:23>

-- Output to stderr:
/Users/mb013619/dev/neovim/src/nvim/screen.c:3364:28: runtime error: index 301 out of bounds for type 'char_u [300]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/mb013619/dev/neovim/src/nvim/screen.c:3364:28 in

CMake Error at /Users/mb013619/dev/neovim/cmake/RunTests.cmake:53 (message):
  Running functional tests failed with error: 1.
```

Test run with the fix to screen.c:
```
[----------] Running tests from test/functional/ui/spell_spec.lua
[ RUN      ] combining long lines #7937 successfully joins lines with "set spell": 64.06 ms OK
[----------] 1 test from test/functional/ui/spell_spec.lua (327.31 ms total)

[----------] Global test environment teardown.
[==========] 1 test from 1 test file ran. (327.43 ms total)
[  PASSED  ] 1 test.
[100%] Built target functionaltest
```

File: test/functional/ui/spell_spec.lua (hunk 2 above)
Comment (brailsmt) [reply test/functional/ui/spell_spec.lua]: It does.  Updating.

File: test/functional/ui/spell_spec.lua (hunk 3 above)
Comment (brailsmt) [reply test/functional/ui/spell_spec.lua]: Not sure how to test highlighting.  I'll try to get it added.

Comment (bfredl) [reply test/functional/ui/spell_spec.lua]: If you use `snapshot_util()` it should give the highlights, then they can be moved to a `screen:set_default_attr_ids({...}}` call. See e g `test/functional/ui/bufhl_spec.lua`

**Comment by brailsmt** [comment]
I'm not sure why the quickbuild status check keeps failing.

**Comment by jamessan** [comment]
QB failure:

```
12:57:22,860 INFO  - not ok 2801 - combining long lines #7937 successfully joins lines with "set spell"
12:57:22,860 INFO  - # test/functional/ui/spell_spec.lua @ 23
12:57:22,860 INFO  - # Failure message: ./test/functional/ui/screen.lua:307: Row 2 did not match.
12:57:22,860 INFO  - # Expected:
12:57:22,860 INFO  - # |^Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor i|
12:57:22,860 INFO  - # |*~ |
12:57:22,860 INFO  - # |~ |
12:57:22,860 INFO  - # | |
12:57:22,860 INFO  - # Actual:
12:57:22,860 INFO  - # |^Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor i|
12:57:22,860 INFO  - # |*ncididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostru|
12:57:22,861 INFO  - # |d exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aut|
12:57:22,861 INFO  - # | |
12:57:22,861 INFO  - #
12:57:22,861 INFO  - # To print the expect() call that would assert the current screen state, use
12:57:22,861 INFO  - # screen:snapshot_util(). In case of non-deterministic failures, use
12:57:22,861 INFO  - # screen:redraw_debug() to show all intermediate screen states.
12:57:22,861 INFO  - # stack traceback:
12:57:22,861 INFO  - #     ./test/functional/ui/screen.lua:307: in function 'wait'
12:57:22,861 INFO  - #     ./test/functional/ui/screen.lua:221: in function 'expect'
12:57:22,861 INFO  - #     test/functional/ui/spell_spec.lua:34: in function <test/functional/ui/spell_spec.lua:23>
12:57:22,861 INFO  - # 
```

**Comment by jamessan** [comment]
CONTRIBUTING.md links to [this description](https://github.com/neovim/neovim/pull/4718#issuecomment-217631350) of how to see why QB fails.

File: test/functional/ui/spell_spec.lua [hunk 4]
```diff
@@ -15,7 +15,6 @@ describe('combining long lines', function()
     screen = Screen.new(80, 4)
     screen:attach()
     screen:set_default_attr_ids( {
-      [0] = {bold=true, foreground=Screen.colors.Blue},
```
Comment (bfredl) [inline test/functional/ui/spell_spec.lua]: Typically we make the window have enough rows so at least one empty line is visible.

**Comment by brailsmt** [comment]
QuickBuild is failing with [this](http://neovim-qb.szakmeister.net/build/16059/tap_report/failed_tests?reportset=functionaltest-macosx-64-debug), which is in test/functional/eval/timer_spec.lua:136.  I am not sure why it my change would cause it to fail.

**Comment by bfredl** [comment]
It is most likely random failure. You can restart the build by force pushing.

**Comment by brailsmt** [comment]
Oops

**Comment by brailsmt** [comment]
Not what I intended.

**Comment by brailsmt** [comment]
Quickbuild seems to fail on the following.  Could this be related to 6744f48d884902483f95c9b89877da36032e2677:
```
17:27:33,617 ERROR - Step 'master>buildall>build-node?testNode=linux-64>build-and-run-tests>build-and-run-tests-parameterized?buildType=Release>configure-neovim-and-build-nvim' is failed: Failed to run command: mkdir -p build/Release && cd build/Release && cmake -G "Unix Makefiles" -DBUSTED_OUTPUT_TYPE=TAP -DCMAKE_BUILD_TYPE=Release -DTRAVIS_CI_BUILD=ON ../.. && make VERBOSE=1 nvim unittest-prereqs functionaltest-prereqs
Command return code: 2
Command error output: In file included from /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/bufhl_defs.h:6:0,
from /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/map.h:9,
from /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/buffer_defs.h:40,
from /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/normal.h:6,
from /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated...
```

**Comment by bfredl** [comment]
No that only modified the bufhl tests, it can't make the bufhl_defs.h file fail...

**Comment by justinmk** [comment]
quickbuild is broken, still looking for time to fix it. just ignore for now.

**Comment by brailsmt** [comment]
What remains to get this merged?  Aside from squashing commits?

**Comment by bfredl** [comment]
If all test pass (except QB) I think it is ready to merge.

**Pull request merged**